### PR TITLE
[WIP] Validation data generation, removal of loss function overrides

### LINKED ISSUE #228: Non-stratified splitting and overwriting of loss function in classification tasks
Hello,

I'd like to report two issues regarding classification tasks in modnet:

First, the loss function passed to `ModnetModel().fit()` is overwritten with `"categorical_crossentropy"` if `val_data` is not None and `self.multi_label=False`:
https://github.com/ppdebreuck/modnet/blob/e14188d3b8a036bba0a1d9c0a3f538dc58c3cd29/modnet/models/vanilla.py#L400-L411
As the `loss=None` case is already handled before in L352-L360 in the preprocessing of the training data, maybe this could be removed here when preprocessing the validation data?

Second, if `nested=False`, both `FitGenetic` and `ModnetModel.fit_preset()` perform a train test split that is not stratified:
https://github.com/ppdebreuck/modnet/blob/e14188d3b8a036bba0a1d9c0a3f538dc58c3cd29/modnet/models/vanilla.py#L580
https://github.com/ppdebreuck/modnet/blob/e14188d3b8a036bba0a1d9c0a3f538dc58c3cd29/modnet/hyper_opt/fit_genetic.py#L458-L462
This is an issue in the case of imbalanced datasets and it would be helpful if the splitting was stratified for classification tasks.

If you are interested, I'm happy to raise a PR with fixes.

### PR DESCRIPTION
**Pull request by kaueltzen**
Resolves #228 . Work in progress.

## Overview
- Replaces `train_test_split` with explicit generation of validation data that is reproducible, shuffled and stratified (if classification tasks are present)
- Removes override of loss function for classification tasks

## TODO
- rework determination of classification targets / most imbalanced classification target (?)
- enable passing of loss function for model evaluation for classification tasks

### CODE REVIEW

**Comment by kaueltzen** [comment]
Hi @ppdebreuck  I have a question regarding this comment:
https://github.com/ppdebreuck/modnet/issues/228#issuecomment-2437540148

Do you think it would make sense to enable the passing of a loss function for classification tasks in `evaluate` of `MODNetModel`?
Or would you prefer keeping -ROC-AUC (e.g., because it's easier to interpret than cross entropy losses)?

File: modnet/models/vanilla.py [hunk 1]
```diff
@@ -1534,3 +1549,46 @@ def validate_model(
...
+):
+    """
+    Generate train and validation data that is shuffled, reproducible and, if classification, stratified.
+    """
+    train_idx, val_idx = generate_shuffled_and_stratified_val_split(
+        y=np.array(y), val_fraction=val_fraction, classification=classification
+    )
+    return (
+        x[train_idx],
+        [t[train_idx] for t in y],
```
Comment (ppdebreuck) [inline modnet/models/vanilla.py:1592]: This doesn't seem correct ?  `y[train_idx]` should work, right?. It would pick the wrong columns now.

**Review by ppdebreuck — CHANGES_REQUESTED** [review]
Thanks @kaueltzen ! Looks good, there just might be something off in the the stratified split, please let me know if you agree.

**Comment by ppdebreuck** [comment]
> Hi @ppdebreuck I have a question regarding this comment: [#228 (comment)](https://github.com/ppdebreuck/modnet/issues/228#issuecomment-2437540148)
> 
> Do you think it would make sense to enable the passing of a loss function for classification tasks in `evaluate` of `MODNetModel`? Or would you prefer keeping -ROC-AUC (e.g., because it's easier to interpret than cross entropy losses)?

No problem with me, as you can put ROC_AUC as default metric to keep current behavior, while adding flexibility if you need to change it :) (I would put it in a separate PR)

File: modnet/models/vanilla.py [hunk 2]
```diff
@@ -1534,3 +1549,46 @@ def validate_model(
...
+    y: np.ndarray, val_fraction: float, classification: bool
+):
+    """
+    Generate train validation split that is shuffled, reproducible and, if classification, stratified.
+    """
+    if classification:
+        if isinstance(y[0][0], list) or isinstance(y[0][0], np.ndarray):
+            ycv = np.argmax(y[0], axis=1)
+        else:
+            ycv = y[0]
```
[the remark this answers is not in the document]
Comment (kaueltzen) [reply modnet/models/vanilla.py:1564]: Hi @ppdebreuck thanks for pointing that out!
It was written for `MODNetModel.fit()` to handle a list of props from `targets_groups`, but outside of `vanilla.py` / `generate_shuffled_and_stratified_val_data` the targets are indeed not correctly handled, I will change it.

File: modnet/models/vanilla.py (hunk 1 above)
Comment (kaueltzen) [reply modnet/models/vanilla.py:1570]: The `y` of `MODNetModel.fit()` and of `generate_shuffled_and_stratified_val_data` has the first two dimensions (n_target_groups, n_samples) while the `y` of `generate_shuffled_and_stratified_val_split` that is used elsewhere (`data.df_targets.values`) has the first 2 dimensions (n_samples, n_targets).

**Review by ppdebreuck — APPROVED** [review]
Thanks for the modifications @kaueltzen, I think this can be merged ?
